cohortextractor/pandas_utils.py

```diff
@@ -0,0 +1,111 @@
+from pathlib import Path
+
+import pandas
+
+
+def dataframe_to_file(df, filename):
+    extension = Path(filename).suffix.lstrip(".")
+    if extension == "csv":
+        df.to_csv(filename, index=False)
+    elif extension == "feather":
+        df.to_feather(filename)
+    elif extension == "dta":
+        df.to_stata(filename, write_index=False)
+    else:
+        raise RuntimeError(f"Unsupported extension: {extension}")
+
+
+def dataframe_from_rows(covariate_definitions, rows):
+    """
+    Create a DataFrame from an iterator of rows
+    """
+    convertors = {
+        name: get_pandas_convertor(funcname, **kwargs)
+        for name, (funcname, kwargs) in covariate_definitions.items()
+    }
+    # `patient_id` isn't included in `covariate_definitions` so we need to add
+    # it manually
+    convertors["patient_id"] = identity
+
+    # First row is expected to be header row
+    headers = next(rows)
+    convertor_funcs = [convertors[column] for column in headers]
+
+    def convert_row(row):
+        return [convertor(value) for (convertor, value) in zip(convertor_funcs, row)]
+
+    data = map(convert_row, rows)
```

Comment:
This is the line that will exhaust all our RAM, right?

Reply:
Nope, next one down! `map` is lazy in py3 but the DataFrame constructor consumes the iterator.

Reply:
In my day, maps weren't lazy, they did what you told 'em, when you told 'em to do it! </grumble>